nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,222 @@
+        instance: SledInstance,
+    ) -> impl Future<Output = ()> + Send + 'static {
+        let instance_id = instance.instance_id();
+        let watcher = self.clone();
+        let opctx = opctx.child(
+            std::iter::once((
+                "instance_id".to_string(),
+                instance_id.to_string(),
+            ))
+            .collect(),
+        );
+        let client = client.clone();
+
+        async move {
+            let InstanceWatcher { datastore, resolver, max_retries } = watcher;
+            slog::trace!(opctx.log, "checking on instance...");
+            let backoff = backoff::retry_policy_internal_service();
+            let mut retries = 0;
+            let rsp = backoff::retry_notify(
+                backoff,
+                || async {
+                    let rsp = client
+                        .instance_get_state(&instance.instance_id())
+                        .await;
+                    match rsp {
+                        Ok(rsp) => Ok(rsp.into_inner()),
+                        Err(e) if retries == max_retries.get() => {
+                            Err(BackoffError::Permanent(e))
+                        }
+                        Err(
+                            e @ ClientError::InvalidRequest(_)
+                            | e @ ClientError::InvalidUpgrade(_)
+                            | e @ ClientError::UnexpectedResponse(_)
+                            | e @ ClientError::PreHookError(_),
+                        ) => Err(BackoffError::Permanent(e)),
+                        Err(e) => Err(BackoffError::transient(e)),
+                    }
+                },
+                |err, duration| {
+                    slog::info!(
+                        opctx.log,
+                        "instance check failed; retrying: {err}";
+                        "duration" => ?duration,
+                        "retries_remaining" => max_retries.get() - retries,
+                    );
+                },
+            )
+            .await;
+            let state = match rsp {
+                Ok(state) => state,
+                Err(error) => {
+                    // TODO(eliza): here is where it gets interesting --- if the
+                    // sled-agent is in a bad state, we need to:
+                    // 1. figure out whether the instance's VMM is reachable directly
+                    // 2. figure out whether we can recover the sled agent?
+                    // 3. if the instances' VMMs are also gone, mark them as
+                    //    "failed"
+                    // 4. this might mean that the whole sled is super gone,
+                    //    figure that out too.
+                    //
+                    // for now though, we'll just log a really big error.
+                    slog::error!(
+                        opctx.log,
+                        "instance seems to be in a bad state: {error}"
+                    );
+                    return;
```

Comment:
For the moment, we are avoiding "automatically marking a whole sled as failed" -- this is an operation that an operator can perform, but we're avoiding automating that behavior.

https://github.com/oxidecomputer/rfd/pull/691#discussion_r1463879601

However, operators will have the ability to mark sled `policy`, which can lead to expungement.

So:
- We could read from the `omicron.public.sled` table
- If the sled is no longer in service, we've decided that the sled **should** be gone
- If that sled still has instances, they **should** fail

As an alternative, when we make the decision to "expunge" a sled, we could automatically mark all instances on that sled "failed" there, too, rather than deferring to this background task.

Reply:
Also, https://github.com/oxidecomputer/omicron/issues/4872 seems relevant!

I *think* that if a sled reboots, it should come back up and explicitly "not return any info about the instance" Nexus queries the Sled Agent (thought it may intermittently "not respond" to Nexus while it's booting up)